server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+        int initialNumBuckets = shardSize * 3/4;
+
+        // Naively use basic linear separation to group the first cacheLimit docs into initialNumBuckets buckets
+        int docsPerBucket = (int) Math.ceil((double) cachedValues.size() / (double) initialNumBuckets);
+
+        int bucketOrd = 0;
+        for(int i = 0; i < indexes.length; i++){
+            // The doc that is currently in bucket indexes[i] is the i'th smallest
+            // We will move it into the the bucket at `bucketOrd`
+
+            int index = indexes[i];
+            double val = cachedValues.get(index);
+
+            mergeMap[index] = bucketOrd;
+            addToCluster(bucketOrd, val);
+
+            if((i + 1) % docsPerBucket == 0){
+                // There are enough docs in this bucket, so we should make a new one
+                bucketOrd += 1;
+            }
+        }
+
+        mergeBuckets(mergeMap, initialNumBuckets);
+        if (deferringCollector != null) {
+            deferringCollector.mergeBuckets(mergeMap);
+        }
+    }
+
+    /**
+     * Returns the ordinal of the bucket whose centroid is closest to <code>val</code>, or -1 if there are no buckets.
+     */
+    private int getNearestBucket(double val) {
+        double smallest_distance = Double.POSITIVE_INFINITY;
+        int nearest_bucket = -1;
+        for(int i = 0; i < clusterCentroids.size(); i++){
+            double current_distance = Math.abs(clusterCentroids.get(i) - val);
+            if(current_distance < smallest_distance){
+                smallest_distance = current_distance;
+                nearest_bucket = i;
+            }
+        }
+        return nearest_bucket;
+    }
+
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) throws IOException {
+        assert bucket == 0;
+
+        if(cachedValues.size() < cacheLimit){
+            bucketCachedDocs();
+        }
+
+        consumeBucketsAndMaybeBreak((int) clusterCentroids.size());
+
+        long[] bucketOrdArray = new long[(int) clusterCentroids.size()];
+        for (int i = 0; i < clusterCentroids.size(); i++) {
+            bucketOrdArray[i] = i;
+        }
+
+        runDeferredCollections(bucketOrdArray);
+
+        List<InternalVariableWidthHistogram.Bucket> buckets = new ArrayList<>((int) clusterCentroids.size());
+        for (long i = 0; i < clusterCentroids.size(); i++) {
+            buckets.add(new InternalVariableWidthHistogram.Bucket(
+                clusterCentroids.get(i),
+                new InternalVariableWidthHistogram.Bucket.BucketBounds(clusterMins.get(i), clusterMaxes.get(i)),
+                bucketDocCount(i),
+                formatter,
+                bucketAggregations(i)));
+        }
+
+        // The contract of the histogram aggregation is that shards must return
+        // buckets ordered by key in ascending order
+        CollectionUtil.introSort(buckets, BucketOrder.key(true).comparator(this));
```

Comment:
I believe that you have set a new contract in your reduce method - they have to be sorted by centroid. They'd come out that way if `clusterCentroids` were sorted which I think is *probably* a good idea.